### ISSUE TITLE
ConfigurationString vs ConfigurationOptions using Redis on Azure

### ISSUE THREAD

**Issue by NTTLuke** [StackExchange/StackExchange.Redis#883]
Hi 
I have a strange behaviour using the Parse method with the same configuration values.

Using Parse with configString , it doesn't work 
Exception Timeout generated during Azure Redis reads
```
var configString = "myDNS:myport,password=mypassword,connectRetry=3,connectTimeout=5000,syncTimeout=5000,defaultDatabase=0,ssl=true,abortConnect=false"
var options = ConfigurationOptions.Parse(configString);

return ConnectionMultiplexer.Connect(options)

```
Using  the same values , but hard coded , it works 
```
return ConnectionMultiplexer.Connect(new ConfigurationOptions
        {
            AbortOnConnectFail = false,
            Ssl = true,
            ConnectRetry = 3,
            ConnectTimeout = 5000,
            SyncTimeout = 5000,
            DefaultDatabase = 0,
            EndPoints = { { myDNS, myPort } },
            Password = myPassword
        });
```
I followed the official documentation for mapping ConfigurationString vs ConfigurationOptions.
https://stackexchange.github.io/StackExchange.Redis/Configuration

Any suggestions ? 
Thank you

**Comment by mgravell** [comment]
Can I just check - on the second ("it works") version, should that be `EndPoints = { { myDNS, myport } },` ? (just trying to make sure I understand before I start poking)

**Comment by NTTLuke** [comment]
oh sorry Marc , my bad during cut and paste.
I updated the code.
Thank you

**Comment by mgravell** [comment]
Question: does your password contain a comma?

**Comment by mgravell** [comment]
Following works for me:

```
        [Fact]
        public void SSLParseViaConfig_Issue883_ConfigObject()
        {
            Skip.IfNoConfig(nameof(TestConfig.Config.AzureCacheServer), TestConfig.Current.AzureCacheServer);
            Skip.IfNoConfig(nameof(TestConfig.Config.AzureCachePassword), TestConfig.Current.AzureCachePassword);

            var options = new ConfigurationOptions
            {
                AbortOnConnectFail = false,
                Ssl = true,
                ConnectRetry = 3,
                ConnectTimeout = 5000,
                SyncTimeout = 5000,
                DefaultDatabase = 0,
                EndPoints = { { TestConfig.Current.AzureCacheServer, 6380 } },
                Password = TestConfig.Current.AzureCachePassword
            };
            using (var conn = ConnectionMultiplexer.Connect(options))
            {
                conn.GetDatabase().Ping();
            }
        }


        [Fact]
        public void SSLParseViaConfig_Issue883_ConfigString()
        {
            Skip.IfNoConfig(nameof(TestConfig.Config.AzureCacheServer), TestConfig.Current.AzureCacheServer);
            Skip.IfNoConfig(nameof(TestConfig.Config.AzureCachePassword), TestConfig.Current.AzureCachePassword);

            var configString = $"{TestConfig.Current.AzureCacheServer}:6380,password={TestConfig.Current.AzureCachePassword},connectRetry=3,connectTimeout=5000,syncTimeout=5000,defaultDatabase=0,ssl=true,abortConnect=false";
            var options = ConfigurationOptions.Parse(configString);
            using (var conn = ConnectionMultiplexer.Connect(options))
            {
                conn.GetDatabase().Ping();
            }
        }
```

I wonder if this is an example of https://github.com/StackExchange/StackExchange.Redis/issues/680

**Comment by NTTLuke** [comment]
Hi Marc 

No commas inside my password.
Thank you for your suggestion and for sample tests.

I ’ll  keep you posted.
Thank you

**Comment by mgravell** [comment]
@NTTLuke  please do, but right now: it seems to work locally - very curious as to what is different; my guess is something gnarly in the password, but (and I cannot stress this enough): please don't send me your azure password - I don't want it and it creates too many issues :)

**Comment by NTTLuke** [comment]
Hi @mgravell 
no password for you LOL

I want to recap the situation:
My problem is not during the connection with Azure Redis , It works properly with connectionString and with ConfigOption. I know because during the startup of my Azure Function I create a subscriber on Redis and I see it with PUBSUB CHANNELS command.
For that reason I think the password is ok ( but correct me if I'm wrong)

The exception I receive is during the operation with LRANGE. 
I investigated about the issue and I found some posts related the right configuration (syncTimeout, connectionRetry and so on) 

If I use the stringConnection I receive TimeoutException during the reads operation using LRANGE.
If I use the same configuration with ConfigOption it works.

I tested on my machine and , for a strange reason , it works.
Deployed on Azure Function ...no way 

At this point I'm seriously thinking it could be an issue related Azure Function 🤔 even if I don't understand how it is possible ( and how it is possible no one has reported the same issue)

Sorry if my first post has created misunderstanding.
I will continue to investigate and I'll keep you posted.

Meanwhile , thank you for your time.

**Comment by NickCraver** [comment]
@NTTLuke any news here?

**Comment by NTTLuke** [comment]
Hi @NickCraver  
I'm just come back from vacation.
I will resume my tests starting from Friday.

I'll keep you posted!
Thank you

**Comment by NTTLuke** [comment]
Hi everyone,
my last tests have the same result.
Since I needed to go in production with the POC , I decided to add preprocessor directives as long as I do not have time to verify better.

One test I had in mind was to check if the problem can be related to Azure Function v2 which are still in beta.

thank you

**Comment by mgravell** [comment]
I simply can't repro. I'd be more than happy to investigate, but I have not managed to repro this. Do you have a string that repros this? (obviously don't post real passwords to github)

**Comment by NTTLuke** [comment]
Hi @mgravell 

I used Azure function v2 (preview) and inside app config the I have a key (RedisUri) with this string 

`"myDNS:myport,password=mypassword,connectRetry=3,connectTimeout=5000,syncTimeout=5000,defaultDatabase=0,ssl=true,abortConnect=false"`


thanks for your support 🙏

**Comment by mgravell** [comment]
Yay, I was reading this one the other day and hoping you'd come back. I'll give this another try on Monday. I'll be honest - this is genuinely one of the most intriguing error reports I've seen in a while. Maybe I need to get out more.

**Comment by mgravell** [comment]
Well; I've added a test that uses reflection to manually compare and contrast: they come out identical. In case this wasn't where it was happening, I traced through the `Connect...` code and ripped out the code that deals with preparing the options into a testable method that I could use... still no difference.

Basically, I cannot detect **any** difference between the two versions. I do not think that the parse code is the thing that is changing the behaviour, unless this is due to some fix that has happened between (whatever version you're using) and what I have locally (the v2 "about to ship" code). My results locally:

```
Compiled under .NET 4.6.2
Running on: Microsoft Windows 10.0.17134 
13:39:02.327: computed: myDNS:883,syncTimeout=5000,connectTimeout=5000,password=mypassword,ssl=True,abortConnect=False,connectRetry=3,defaultDatabase=0
13:39:02.327: Checking endpoints...
13:39:02.355: Host: myDNS vs myDNS
13:39:02.357: Port: 883 vs 883
13:39:02.357: AddressFamily: Unspecified vs Unspecified
13:39:02.357: Comparing 30 fields...
13:39:02.358: <ChannelPrefix>k__BackingField: (null) vs (null)
13:39:02.368: <ClientName>k__BackingField: (null) vs (null)
13:39:02.368: <DefaultDatabase>k__BackingField: 0 vs 0
13:39:02.368: <EndPoints>k__BackingField: StackExchange.Redis.EndPointCollection vs StackExchange.Redis.EndPointCollection
13:39:02.369: <Password>k__BackingField: mypassword vs mypassword
13:39:02.369: <ServiceName>k__BackingField: (null) vs (null)
13:39:02.369: <SocketManager>k__BackingField: (null) vs (null)
13:39:02.369: <SslProtocols>k__BackingField: (null) vs (null)
13:39:02.369: abortOnConnectFail: False vs False
13:39:02.369: allowAdmin: (null) vs (null)
13:39:02.369: asyncTimeout: (null) vs (null)
13:39:02.369: CertificateSelection: (null) vs (null)
13:39:02.369: CertificateValidation: (null) vs (null)
13:39:02.369: commandMap: (null) vs (null)
13:39:02.369: configChannel: (null) vs (null)
13:39:02.369: configCheckSeconds: (null) vs (null)
13:39:02.369: connectRetry: 3 vs 3
13:39:02.369: connectTimeout: 5000 vs 5000
 13:39:02.369: defaultVersion: (null) vs (null)
13:39:02.369: highPrioritySocketThreads: (null) vs (null)
13:39:02.369: keepAlive: (null) vs (null)
13:39:02.369: proxy: (null) vs (null)
13:39:02.369: reconnectRetryPolicy: (null) vs (null)
13:39:02.369: resolveDns: (null) vs (null)
13:39:02.369: responseTimeout: (null) vs (null)
13:39:02.369: ssl: True vs True
13:39:02.369: sslHost: (null) vs (null)
13:39:02.369: syncTimeout: 5000 vs 5000
13:39:02.369: tieBreaker: (null) vs (null)
13:39:02.369: writeBuffer: (null) vs (null)
```

Question: what exact environment are you in, so I can try to repro there? What OS and .NET framework / version?

However, I increasingly suspect that we're chasing the wrong thing here, and that maybe what you're seeing is bad cold-start performance from the cloud provider. Hypothesis: it times out due to slow start, so you change something and retry and it works - but maybe it isn't the change that made it work, but just the fact that enough time has passed for the service to be usable.

**Comment by mgravell** [comment]
one other possibility: what *locale*? perhaps the integer parse code is failing for a locale reason? again: just a thought

**Comment by mgravell** [comment]
ooh, I have found *one* `int.TryParse` that doesn't use explicit culture... checking it

**Comment by mgravell** [comment]
```
13:56:57.372: Checking 843 cultures...
```

and... all 843 are "green" and parse correctly :(

**Comment by mgravell** [comment]
but: I'm still interested in your local locale, so I can check...

**Comment by NTTLuke** [comment]
Hi @mgravell you are not a man you are a superhero!!! 
Sorry for late response .
I has been moved to other stuffs , so at time being my colleague is still working with my  patch mentioned ( since it is just a poc) 
I will take a look in the weekend in my free time just for my curiosity ( probably I need to get out too .. my wife says I’m too white🙈) 
I will keep you posted 
In the meantime I want to say thank you for your support. 
Really you are a super hero !

**Comment by mgravell** [comment]
closing as unable to repro

**Comment by azarc3** [comment]
Suggest re-opening this as I ran into the same issue this morning (2020.08.24) while adding the **sslProtocols** parameter. 

**TL;DR Version:**
Looks like the parameter parser doesn't handle the final parameter key/value set correctly.

**Longer Version:**
I thought the issue was with my sslProtocols value and I sat there spinning my wheels for about an hour while I tried different combinations. Finally, I tried moving the parameter "earlier" in the string... and then a different parameter started failing. The only commonality was that the failing parameter was always at the end of the connection string.

On a hunch I tried adding a comma after the final parameter and Wah-lah! Everything works as expected!

Suggest updating the **ConfigurationOptions.Parse**() method to correctly handle the final parameter of a passed-in configuration string.

**Comment by mgravell** [comment]
Happy to take another look, but if you had a failing example: that's be great

**Comment by azarc3** [comment]
> Happy to take another look, but if you had a failing example: that's be great

Sure; the string that had the problem (scary stuff redacted) is:
```
myDNS:myPort,password=myPassword,connectRetry=3,connectTimeout=15000,syncTimeout=15000,defaultDatabase=0,abortConnect=false,ssl=true,sslProtocols=Tls12
```
I checked, and the following simplified string fails as well:
```
"myDNS:myPort,password=myPassword,abortConnect=false,ssl=true,sslProtocols=Tls12"
```

Here's the Errror Message and redacted Stack Trace it produces:

**Error Message:**
```
ArgumentOutOfRangeException -	Specified argument was out of the range of valid values.
Parameter name: Keyword 'sslProtocols' requires an SslProtocol value (multiple values separated by '|').
```

**Stack Trace:**
```
   at StackExchange.Redis.ConfigurationOptions.OptionKeys.ParseSslProtocols(String key, String value)
   at StackExchange.Redis.ConfigurationOptions.DoParse(String configuration, Boolean ignoreUnknown)
   at StackExchange.Redis.ConfigurationOptions.Parse(String configuration)
   at MyNamespace.Cache.ParseConnectionOptions(String connectionOptions) in C:\MyPath\MyNamespace.Cache.Redis\Cache.cs:line 343
   at GSP.Cache.Redis.Cache.Init(String cacheName, String connectionOptions, Boolean readOnlyStorage) in C:\MyPath\MyNamespace.Cache.Redis\Cache.cs:line 112
```

Simply adding a trailing comma delimiter to the end of the string gets me past this issue.

**Comment by mgravell** [comment]
Awesome, thanks. Will investigate.

**Comment by mgravell** [comment]
Cannot repro; the following pass just fine (😕):

``` c#
[Theory]
[InlineData("myDNS:myPort,password=myPassword,connectRetry=3,connectTimeout=15000,syncTimeout=15000,defaultDatabase=0,abortConnect=false,ssl=true,sslProtocols=Tls12", SslProtocols.Tls12)]
[InlineData("myDNS:myPort,password=myPassword,abortConnect=false,ssl=true,sslProtocols=Tls12", SslProtocols.Tls12)]
#pragma warning disable CS0618 // obsolete
[InlineData("myDNS:myPort,password=myPassword,abortConnect=false,ssl=true,sslProtocols=Ssl3", SslProtocols.Ssl3)]
#pragma warning restore CS0618 // obsolete
public void ParseTlsWithoutTrailingComma(string configString, SslProtocols expected)
{
    var config = ConfigurationOptions.Parse(configString);
    Assert.Equal(expected, config.SslProtocols);
}
```

**Comment by mgravell** [comment]
what I *am* going to do, however, is improve the error logging in these cases - to say what value it is trying to parse. My suspicion is that in your specific case (@azarc3) something in the code we can't see is concatenating something else onto the string, making it illegal. The proposed change will make that more obvious.
